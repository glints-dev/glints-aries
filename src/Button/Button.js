/* @flow */

import React, { Fragment } from 'react';
import styled, {css} from 'styled-components';
import CtaButton from './CtaButton';
import JobCardButton from './JobCardButton';
import SecondaryButton from './SecondaryButton';

type ButtonVariant = 'default' | 'cta';
type ButtonType = 'primary' | 'secondary' | 'info'

type Props = {
	disabled: boolean,
	variant?: ButtonVariant,
	type?: ButtonType,
}

type HandlerProps = {
	disabled: boolean,
	variant?: ButtonVariant,
	type?: ButtonType
};

const buttonStyleHandler = (props: HandlerProps) => {
  const styles: Object = {};

  styles.width = props.block || props.position ? '100%' : 'auto';
  styles.borderRadius = props.round ? '8rem' : '0rem';

  switch (props.variant) {
    case 'cta':
      styles.border = 'solid 0.1rem rgba(0, 0, 0, 0.24)';
      styles.borderBottomColor = 'rgba(0, 0, 0, 0.32)';
      styles.backgroundColor = '#3d70b2';
      styles.mouseEventBackgroundColor = '#30588c';
      break;
    default:
      styles.border = 'solid 0.1rem rgba(0, 0, 0, 0.24)';
      styles.borderBottomColor = 'rgba(0, 0, 0, 0.32)';
      styles.backgroundColor = '#fff';
      styles.mouseEventBackgroundColor = '#d2d2d2';
	}
	
	switch (props.type) {
		case 'primary':
			styles.backgroundColor = 'red';
			break;
		default:
			styles.backgroundColor = '#333';
	}

  styles.cursor = props.disabled ? 'not-allowed' : 'pointer';
  styles.opacity = props.disabled ? '0.72' : '1';
  styles.mouseEventBackgroundColor = props.disabled
    ? styles.backgroundColor
		: styles.mouseEventBackgroundColor;
		
	console.log(styles);

  return styles;
};

const ButtonWrapper = styled.button`
	appearance: none;
	background-color: ${props => buttonStyleHandler(props).backgroundColor};
`

const Button = (props: Props) => {
	const {
		variant,
		type,
		...defaultButtonProps
	} = props;

	return (
		<ButtonWrapper
			variant={variant}
			type={type}
			{...defaultButtonProps}
		>
			{props.children}
		</ButtonWrapper>
	);
}

export default Button;