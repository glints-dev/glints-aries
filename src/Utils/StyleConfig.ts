export const Variant = {
  PRIMARY: 'primary',
  SECONDARY: 'secondary',
  GHOST: 'ghost',
  LINK: 'link',
};

export const TabVariant = {
  HORIZONTAL: 'horizontal',
  VERTICAL: 'vertical',
};

export const PsychedelicTheme = {
  BLUE_DOMINANT: 'blue-dominant',
  BLUE_DEFAULT: 'blue-default',
  RED_DOMINANT: 'red-dominant',
  RED_DEFAULT: 'red-default',
  YELLOW_DOMINANT: 'yellow-dominant',
  YELLOW_DEFAULT: 'yellow-default',
};

export const Theme = {
  RED: 'red',
  YELLOW: 'yellow',
  BLUE: 'blue',
  BLUE_RED: 'blue-red',
  WHITE: 'white',
  GREY: 'grey',
};

/* Screen size library of Glints
 * Usage:
 * @media ${Device.mobileL} { ... }
 */

export const ScreenSize = {
  mobileS: 320,
  mobileM: 480,
  mobileL: 640,
  tablet: 768,
  desktopS: 1024,
  desktopM: 1260,
  desktopL: 1440,
};

export const Device = {
  mobileS: `(max-width: ${ScreenSize.mobileS}px)`,
  mobileM: `(max-width: ${ScreenSize.mobileM}px)`,
  mobileL: `(max-width: ${ScreenSize.mobileL}px)`,
  tablet: `(max-width: ${ScreenSize.tablet}px)`,
  desktopS: `(max-width: ${ScreenSize.desktopS}px)`,
  desktopM: `(max-width: ${ScreenSize.desktopM}px)`,
  desktopL: `(max-width: ${ScreenSize.desktopL}px)`,
};
