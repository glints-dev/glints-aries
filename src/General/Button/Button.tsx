import React, { FC, ReactNode } from 'react';

import SolidButton from './SolidButton';
import SolidShadowButton from './SolidShadowButton';
import GhostButton from './GhostButton';
import LinkButton from './LinkButton';

import { ButtonVariant, ButtonTheme } from '../../Utils/StyleConfig';
import {
  StartIconContainer,
  EndIconContainer,
  WhiteGreyBtn,
} from './ButtonStyle';

const SOLID_BUTTON_THEME_MAP = {
  [ButtonVariant.SOLID_WHITE]: ButtonTheme.WHITE,
  [ButtonVariant.SOLID_BLUE]: ButtonTheme.BLUE,
};

export const transformVariant = (variant: string, theme?: string) => {
  if (theme) {
    console.warn(
      `Warning: Button's theme prop is deprecated and will be removed in a future release.\nPlease use the variant prop instead.`
    );
  }

  if (
    (variant === ButtonVariant.SOLID_WHITE || variant === 'default') &&
    theme &&
    theme !== ButtonTheme.WHITE
  ) {
    console.warn(
      `Warning: Button's 'default' variant is deprecated and will be removed in a future release.\nPlease use the 'solid-blue' or 'solid-white' variant instead.`
    );
    return ButtonVariant.SOLID_BLUE;
  } else if (variant === 'primary') {
    console.warn(
      `Warning: Button's 'primary' variant is deprecated and will be removed in a future release.\nPlease use the 'yellow' variant instead.`
    );
    return ButtonVariant.YELLOW;
  } else if (Object.values(ButtonVariant).includes(variant)) {
    return variant;
  }
  return ButtonVariant.SOLID_WHITE;
};

export const Button: FC<Props> = ({
  children,
  block = false,
  className,
  disabled = false,
  onClick,
  small = false,
  variant = ButtonVariant.SOLID_WHITE,
  startIcon,
  endIcon,
  theme,
  ...defaultProps
}) => {
  const content = (
    <>
      {startIcon && <StartIconContainer>{startIcon}</StartIconContainer>}
      {children}
      {endIcon && <EndIconContainer>{endIcon}</EndIconContainer>}
    </>
  );

  const transformedVariant = transformVariant(variant, theme);

  switch (transformedVariant) {
    case ButtonVariant.YELLOW:
      return (
        <SolidShadowButton
          className={className}
          disabled={disabled}
          onClick={onClick}
          block={block}
          small={small}
          {...defaultProps}
        >
          {content}
        </SolidShadowButton>
      );
    case ButtonVariant.GHOST:
      return (
        <GhostButton
          className={className}
          disabled={disabled}
          onClick={onClick}
          block={block}
          small={small}
          {...defaultProps}
        >
          {content}
        </GhostButton>
      );
    case ButtonVariant.WHITE_GREY:
      return (
        <WhiteGreyBtn
          className={className}
          disabled={disabled}
          onClick={onClick}
          block={block}
          small={small}
          {...defaultProps}
        >
          {content}
        </WhiteGreyBtn>
      );
    case ButtonVariant.LINK:
      return (
        <LinkButton
          className={className}
          disabled={disabled}
          onClick={onClick}
          block={block}
          {...defaultProps}
        >
          {content}
        </LinkButton>
      );
    default:
      return (
        <SolidButton
          theme={SOLID_BUTTON_THEME_MAP[transformedVariant]}
          className={className}
          disabled={disabled}
          onClick={onClick}
          block={block}
          small={small}
          {...defaultProps}
        >
          {content}
        </SolidButton>
      );
  }
};

export type ButtonVariantType =
  | 'solid-white'
  | 'solid-blue'
  | 'yellow'
  | 'ghost'
  | 'white-grey'
  | 'link';

export interface Props extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  /** Make the button fit to its parent width. */
  block?: boolean;
  className?: string;
  /** Sets Button to disable state. */
  disabled?: boolean;
  onClick?(e: React.MouseEvent<HTMLButtonElement, MouseEvent>): void;
  /** Sets Button to small version. */
  small?: boolean;
  variant?: ButtonVariantType;
  /** Changes the tag with which the button will render. */
  tag?: React.ElementType;
  /** Element placed before the children. */
  startIcon?: React.ReactNode;
  /** Element placed after the children. */
  endIcon?: React.ReactNode;
  /** Deprecated! */
  theme?: string;
  forwardedRef?: React.RefObject<HTMLButtonElement>;
}

const forwardRef = (props: Props, ref: React.RefObject<HTMLButtonElement>) => (
  <Button {...props} forwardedRef={ref} />
);

forwardRef.displayName = Button.name;

export default React.forwardRef(forwardRef);
