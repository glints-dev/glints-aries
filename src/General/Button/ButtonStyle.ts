import styled from 'styled-components';
import { PrimaryColor, SecondaryColor, Greyscale } from '../../Utils/Colors';
import { ButtonTheme } from '../../Utils/StyleConfig';

const generalButtonPadding = [20, 40];
const smallButtonPadding = [5, 10];

interface ButtonProps {
  small?: boolean;
}

const Button = styled.button<ButtonProps>`
  display: flex;
  position: relative;
  align-items: center;
  justify-content: center;
  border: none;
  text-transform: uppercase;
  outline: none;
  cursor: pointer;
  font-weight: bold;
  font-size: ${({ small }) => (small ? '14px' : '20px')};
  line-height: 1.5;
  padding: ${({ small }) =>
    small
      ? `${smallButtonPadding[0]}px ${smallButtonPadding[1]}px`
      : `${generalButtonPadding[0]}px ${generalButtonPadding[1]}px`};
  background: transparent;
  transition: all 0.2s;
  text-decoration: none;

  ${({ disabled }) => {
    if (disabled) {
      return `
        &:hover {
          cursor: not-allowed;
        }
      `;
    }
  }}
`;

/*
 * Default Button
 */

interface SolidBtnContainerProps {
  block?: boolean;
  disabled?: boolean;
}

interface SolidBtnProps {
  block?: boolean;
}

export const SolidBtn = styled(Button)<SolidBtnProps>`
  width: ${({ block }) => block && '100%'};

  &:active {
    background-color: ${Greyscale.black};
    color: ${Greyscale.white};
  }

  ${props => {
    switch (props.theme) {
      case `${ButtonTheme.BLUE}`:
        return `
          background-color: ${SecondaryColor.actionblue};
          color: ${Greyscale.white};

          &:hover {
            color: ${Greyscale.white};
          }
        `;
      default:
        return `
          background-color: ${Greyscale.white};
          color: ${SecondaryColor.actionblue};

          &:hover {
            color: ${SecondaryColor.actionblue};
          }
        `;
    }
  }}

  ${({ disabled }) => {
    if (disabled) {
      return `
        background-color: ${Greyscale.lightgrey};
        color: ${Greyscale.white};
      `;
    }
  }}

  &:hover {
    text-decoration: none;
    ${({ disabled }) => {
      if (disabled) {
        return `
          background-color: ${Greyscale.lightgrey};
          color: ${Greyscale.white};
        `;
      }
    }}
  }
`;

export const SolidBtnContainer = styled.div<SolidBtnContainerProps>`
  position: relative;
  display: ${({ block }) => (block ? 'flex' : 'inline-flex')};
  z-index: 1;

  ${({ disabled }) => {
    if (!disabled) {
      return `
      &:active {
        background: ${Greyscale.black};
        color: ${Greyscale.white};
        transform: translate3d(2px, 2px, 0);
        transition: all .2s;
      }

      &:after {
        content: '';
        opacity: 0;
        position: absolute;
        width: 100%;
        height: 100%;
        background: repeating-linear-gradient(-45deg, transparent, transparent 3px, rgba(0, 0, 0, 0.5) 5px);
        top: 5px;
        left: 5px;
        z-index: -1;
        transition: all .2s;
        cursor: pointer;
      }

      &:hover:after {
        opacity: 1;
        transition: all .2s;
      }

      &:active:after {
        transform: translate3d(-5px, -5px, 0);
        transition: all .2s;
      }
      `;
    }
  }}

  &:hover {
    ${SolidBtn} {
      ${props => {
        if (props.disabled) {
          return `
            background-color: none;
          `;
        }
      }}
    }
  }

  &:active {
    ${SolidBtn} {
      ${props => {
        if (props.disabled) {
          return `
            background-color: none;
          `;
        }
        return `
          background-color: ${Greyscale.black};
          color: ${Greyscale.white};
        `;
      }}
    }
  }
`;

/*
 * SolidShadow Button
 */

interface SolidShadowContainerProps {
  block?: boolean;
  disabled?: boolean;
}

interface SolidShadowBtnProps {
  block?: boolean;
}

export const SolidShadowBtn = styled(Button)<SolidShadowBtnProps>`
  transition: all 0.2s;
  cursor: pointer;
  position: relative;
  width: 100%;

  &:hover {
    text-decoration: none;
  }

  ${props => {
    if (props.disabled) {
      return `
        background-color: ${Greyscale.lightgrey};
        color: ${Greyscale.white};
      `;
    }
    return `
      background-color: ${PrimaryColor.glintsyellow};
      color: ${Greyscale.black};
    `;
  }}
`;

export const SolidShadowContainer = styled.div<SolidShadowContainerProps>`
  position: relative;
  display: ${({ block }) => (block ? 'flex' : 'inline-flex')};
  z-index: 1;
  cursor: pointer;

  &:hover {
    ${props => {
      if (props.disabled) {
        return 'transform: none';
      }
      return `
        transform: translate(2px, 2px);
        transition: all .2s;
      `;
    }}
  }

  &:active {
    ${props => {
      if (!props.disabled) {
        return `
          transform: translate(3px, 3px);
          transition: all .2s;
          color: ${Greyscale.white};
        `;
      }

      return 'transform: none';
    }}

    ${SolidShadowBtn} {
      ${props => {
        if (props.disabled) {
          return `
            background-color: none;
          `;
        }
        return `
          background-color: ${Greyscale.black};
          color: ${Greyscale.white};
        `;
      }}
    }
  }

  &:before {
    content: '';
    width: 100%;
    height: 100%;
    position: absolute;
    top: 5px;
    left: 5px;
    transition: all 0.2s;

    ${props => {
      if (props.disabled) {
        return 'background-color: none';
      }
      return `background-color: ${PrimaryColor.glintsred};`;
    }}
  }

  &:hover:before {
    ${props => {
      if (props.disabled) {
        return `
          cursor: default;
          background-color: none;
        `;
      }
      return `
      background: ${Greyscale.black};
      transform: translate(-2px, -2px);
      transition: all .2s;
    `;
    }}
  }

  &:active:before {
    ${props => {
      if (!props.disabled) {
        return `
        transform: translate(-5px, -5px);
        transition: all .2s;
      `;
      }

      return false;
    }}
  }
`;

/*
 * Ghost Button
 */

export const GhostBtn = styled(Button)<GhostBtnProps>`
  transition: background-color 0.5s;
  width: ${({ block }) => block && '100%'};
  background: ${Greyscale.white};
  padding: ${({ small }) =>
    small
      ? `${smallButtonPadding[0] - 2}px ${smallButtonPadding[1] - 2}px`
      : `${generalButtonPadding[0] - 2}px ${generalButtonPadding[1] - 2}px`};
  border: 2px solid ${SecondaryColor.actionblue};
  color: ${SecondaryColor.actionblue};

  &:hover {
    transition: background-color 0.5s;
    text-decoration: none;

    ${({ disabled }) => {
      if (!disabled) {
        return `
          background-color: ${SecondaryColor.actionblue};
          color: ${Greyscale.white};
        `;
      }
    }}
  }

  ${({ disabled }) => {
    if (disabled) {
      return `
        background-color: ${SecondaryColor.lightgrey};
        color: ${SecondaryColor.white};
        border: none;
      `;
    }
  }}
`;

interface GhostBtnProps {
  block?: boolean;
}

export const GhostBtnContainer = styled.div<GhostBtnContainerProps>`
  position: relative;
  display: ${({ block }) => (block ? 'flex' : 'inline-flex')};
  z-index: 1;

  ${({ disabled }) => {
    if (!disabled) {
      return `
        &:active {
          transition: background-color .5s; 
          transform: translate3d(2px, 2px, 0);
          transition: all .2s;
        }

        &:after {
          content: '';
          opacity: 0;
          position: absolute;
          width: 100%;
          height: 100%;
          background: repeating-linear-gradient(-45deg, transparent, transparent 3px, rgba(0, 0, 0, 0.5) 5px);
          top: 5px;
          left: 5px;
          z-index: -1;
          transition: all .2s;
          cursor: pointer;
        }

        &:hover:after {
          opacity: 1;
          transition: all .2s;
        }

        &:active:after {
          transform: translate3d(-5px, -5px, 0);
          transition: all .2s;
        }
        `;
    }
  }}

  ${GhostBtn} {
    ${({ disabled }) => {
      if (!disabled) {
        return `
        &:active {
          background-color: ${Greyscale.black};
          color: ${Greyscale.white};
          border: 2px solid ${Greyscale.black};
        }
      `;
      }
    }}
  }
`;

interface GhostBtnContainerProps {
  block?: boolean;
  disabled?: boolean;
}

/*
 * White-Grey Button
 */
export const WhiteGreyBtn = styled(Button)<WhiteGreyBtnProps>`
  width: ${({ block }) => block && '100%'};

  ${({ disabled }) => {
    if (!disabled) {
      return `
        background-color: ${Greyscale.white};
        color: ${SecondaryColor.actionblue};

        &:hover {
          background-color: ${Greyscale.softgrey};
          color: ${SecondaryColor.actionblue};
        }

        &:active {
          background-color: ${Greyscale.black};
          color: ${Greyscale.white};
        }
      `;
    }
    return `
      background-color: ${Greyscale.lightgrey};
      color: ${Greyscale.white};
    `;
  }}
`;

interface WhiteGreyBtnProps {
  block?: boolean;
}

/*
 * Link Button
 */

export const LinkBtn = styled(Button)<LinkBtnProps>`
  font-weight: normal;
  text-transform: inherit;
  width: ${({ block }) => block && '100%'};
  padding: 0;
  justify-content: flex-start;
  color: ${SecondaryColor.actionblue};

  ${({ disabled }) => {
    if (disabled) {
      return `
        color: ${Greyscale.grey};
      `;
    } else {
      return `
        &:hover {
          color: ${SecondaryColor.darkblue};
          text-decoration: none;
        }
      
        &:active {
          color: ${Greyscale.black};
        }
      `;
    }
  }}
`;

interface LinkBtnProps {
  block?: boolean;
  disabled?: boolean;
}

export const StartIconContainer = styled.span`
  display: inline-flex;
  margin-right: 10px;
`;

export const EndIconContainer = styled.span`
  display: inline-flex;
  margin-left: 10px;
`;
