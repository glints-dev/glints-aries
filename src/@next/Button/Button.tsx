import React, { MouseEvent } from 'react';
import { Spinner } from '../Spinner/Spinner';
import { Typography } from '../Typography';
import { StyledButton } from './ButtonStyle';

export type ButtonSize = 'default' | 'slim' | 'large';
export type IconPosition = 'left' | 'right';
export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  size?: ButtonSize;
  children?: React.ReactNode;
  /** Indicate if the button is currently selected  */
  active?: boolean;
  fullWidth?: boolean;
  icon?: React.ReactNode;
  iconPosition?: IconPosition;
  loading?: boolean;
}

export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  function Button(
    {
      children,
      icon,
      iconPosition = 'left',
      active,
      fullWidth,
      loading,
      ...otherProps
    }: ButtonProps,
    ref
  ) {
    const content =
      typeof children === 'string' ? (
        <Typography variant="button"> {children}</Typography>
      ) : (
        children
      );

    const renderIcon = (position: IconPosition) => {
      if (icon && position === iconPosition) {
        return icon;
      }

      return null;
    };

    return (
      <StyledButton
        ref={ref}
        data-active={active}
        data-full-width={fullWidth}
        data-loading={loading}
        data-icon={!!icon}
        {...otherProps}
        onMouseUp={(e: MouseEvent<HTMLButtonElement>) => e.currentTarget.blur()}
      >
        {loading && <Spinner />}
        {renderIcon('left')}
        {content}
        {renderIcon('right')}
      </StyledButton>
    );
  }
);
