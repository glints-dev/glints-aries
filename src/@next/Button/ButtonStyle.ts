import styled from 'styled-components';
import * as Breakpoints from '../utilities/breakpoints';
import { borderRadius4 } from '../utilities/borderRadius';
import { Neutral } from '../utilities/colors';
import {
  button,
  buttonFocused,
  insetButtonPressed,
} from '../utilities/dropShadow';
import {
  space12,
  space16,
  space24,
  space4,
  space8,
  space2,
} from '../utilities/spacing';
import { ButtonProps, ButtonSize } from './Button';

const defaultButtonSizeStyle = `
  min-width: 75px;
  height: 36px;
  padding: ${space8} ${space16};

  @media (max-width: ${Breakpoints.large}) {
    min-width: 71px;
  }
`;

const largeButtonSizeStyle = `
  min-width: 91px;
  height: 44px;
  padding: ${space12} ${space24};

  @media (max-width: ${Breakpoints.large}) {
    min-width: 87px;
  }
`;

const slimButtonSizeStyle = `
  min-width: 67px;
  height: 32px;
  padding: ${space4} ${space12};

  @media (max-width: ${Breakpoints.large}) {
    min-width: 63px;
  }
`;

const verySlimButtonSizeStyle = `
  padding: 5px ${space2};

  @media (max-width: ${Breakpoints.large}) {
    padding: 0px 0px;
  }

  p {
    margin-block-start: 0px;
    margin-block-end: 0px;
  }
`;

const buttonSizeStyleMap: Record<ButtonSize, string> = {
  default: defaultButtonSizeStyle,
  slim: slimButtonSizeStyle,
  large: largeButtonSizeStyle,
  verySlim: verySlimButtonSizeStyle,
};

export const StyledButton = styled.button<ButtonProps>`
  display: flex;
  gap: 8px;
  border-width: 1px;
  border-style: solid;
  border-radius: ${borderRadius4};
  flex-direction: row;
  align-items: center;
  justify-content: center;
  flex: none;
  order: 0;
  flex-grow: 0;
  color: ${Neutral.B18};
  background: ${Neutral.B100};
  border-color: ${Neutral.B68};
  box-shadow: ${button};

  &:hover {
    cursor: pointer;
    background: ${Neutral.B99};
  }

  &:focus {
    outline: none;
    border-color: ${Neutral.B68};
    box-shadow: ${buttonFocused};
  }

  &:active {
    background: ${Neutral.B95};
    border: 1px solid ${Neutral.B68};
    box-shadow: ${insetButtonPressed};
  }

  &:active > svg {
    fill: ${Neutral.B40};
  }

  &:disabled {
    color: ${Neutral.B68};
    background: ${Neutral.B95};
    pointer-events: none;
  }

  &:disabled svg {
    fill: ${Neutral.B85};
  }

  ${({ size = 'default' }: ButtonProps) => {
    const sizeStyle = buttonSizeStyleMap[size];
    if (size && !sizeStyle) {
      console.warn(`${size} is not a valid ButtonSize, default will be used`);
    }
    return buttonSizeStyleMap[size] || defaultButtonSizeStyle;
  }}

  svg {
    width: 20px;
    height: 20px;
    fill: ${Neutral.B40};
  }

  .spinner-container {
    position: absolute;
  }

  &[data-active='true'] {
    color: ${Neutral.B100};
    background: ${Neutral.B40};
  }

  &[data-full-width='true'] {
    width: 100%;
  }

  &[data-icon='true'] {
    gap: 4px;
  }

  &[data-active='true'] svg {
    fill: ${Neutral.B100};
  }

  &[data-loading='true'] {
    background: ${Neutral.B99};
    color: transparent;
  }

  &[data-loading='true'] svg {
    fill: transparent;
  }

  &[data-loading='true'] .spinner-container svg {
    fill: ${Neutral.B85};
    position: static;
  }
`;
