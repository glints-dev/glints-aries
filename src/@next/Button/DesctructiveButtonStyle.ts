import styled from 'styled-components';
import { Neutral, Red } from '../utilities/colors';
import { insetButtonPressed } from '../utilities/dropShadow';
import { Button } from './Button';

export const DestructiveButton = styled(Button)`
  color: ${Neutral.B100};
  background: ${Red.B93};
  border: none;

  &:hover {
    background: ${Red.B74};
    box-shadow:
      0px 1px 0px rgba(0, 0, 0, 0.08),
      inset 0px -1px 0px rgba(0, 0, 0, 0.2);
  }

  &:focus {
    box-shadow: 0px 0px 0px 2px rgba(69, 143, 255, 1);
  }

  &:active {
    background: ${Red.B65};
    box-shadow: ${insetButtonPressed};
  }

  &[data-loading='true'] {
    background: ${Neutral.B95};
  }
`;
