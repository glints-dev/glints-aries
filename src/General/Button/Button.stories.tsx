import React from 'react';
import { Story, Meta } from '@storybook/react';

import { Button, Props as ButtonProps } from './Button';
import {
  ArrowBackDoubleIcon,
  ArrowBackIcon,
  ArrowNextDoubleIcon,
  ArrowNextIcon,
  ViewIcon,
} from '../Icon/components';
import styled from 'styled-components';

const ButtonStack = styled('div')`
  display: flex;
  flex-direction: column;
  align-items: flex-start;

  > :not(:last-child) {
    margin-bottom: 10px;
  }
`;

export default {
  title: 'General/Button',
  component: Button,
  argTypes: {
    startIcon: {
      control: {
        type: 'select',
        options: {
          ArrowBackIcon: <ArrowBackIcon />,
          ArrowBackDoubleIcon: <ArrowBackDoubleIcon />,
        },
      },
    },
    endIcon: {
      control: {
        type: 'select',
        options: {
          ArrowBackIcon: <ArrowNextIcon />,
          ArrowBackDoubleIcon: <ArrowNextDoubleIcon />,
        },
      },
    },
  },
} as Meta;

const Template: Story<ButtonProps> = args => <Button {...args}>Button</Button>;

export const SolidBlue = Template.bind({});
SolidBlue.args = {
  variant: 'solid-blue',
};

export const Ghost = Template.bind({});
Ghost.args = {
  variant: 'ghost',
};

export const Variants: Story<ButtonProps> = () => (
  <ButtonStack>
    <Button variant="solid-white">Solid-white</Button>
    <Button variant="solid-blue">Solid-blue</Button>
    <Button variant="yellow">Yellow</Button>
    <Button variant="ghost">Ghost</Button>
    <Button variant="white-grey">White-grey</Button>
    <Button variant="link">Link</Button>
  </ButtonStack>
);

export const Disabled: Story<ButtonProps> = () => (
  <ButtonStack>
    <Button variant="solid-white" disabled>
      Solid-white
    </Button>
    <Button variant="solid-blue" disabled>
      Solid-blue
    </Button>
    <Button variant="yellow" disabled>
      Yellow
    </Button>
    <Button variant="ghost" disabled>
      Ghost
    </Button>
    <Button variant="white-grey" disabled>
      White-grey
    </Button>
    <Button variant="link" disabled>
      Link
    </Button>
  </ButtonStack>
);

export const Small: Story<ButtonProps> = () => (
  <ButtonStack>
    <Button variant="solid-white" small>
      Solid-white
    </Button>
    <Button variant="solid-blue" small>
      Solid-blue
    </Button>
    <Button variant="yellow" small>
      Yellow
    </Button>
    <Button variant="ghost" small>
      Ghost
    </Button>
    <Button variant="white-grey" small>
      White-grey
    </Button>
    <Button variant="link" small>
      Link
    </Button>
  </ButtonStack>
);

export const Block: Story<ButtonProps> = () => (
  <ButtonStack style={{ alignItems: 'unset' }}>
    <Button variant="solid-white" block>
      Solid-white
    </Button>
    <Button variant="solid-blue" block>
      Solid-blue
    </Button>
    <Button variant="yellow" block>
      Yellow
    </Button>
    <Button variant="ghost" block>
      Ghost
    </Button>
    <Button variant="white-grey" block>
      White-grey
    </Button>
    <Button variant="link" block>
      Link
    </Button>
  </ButtonStack>
);

export const WithIcon: Story<ButtonProps> = () => (
  <ButtonStack>
    <Button
      variant="solid-white"
      startIcon={<ViewIcon />}
      endIcon={<ArrowNextIcon />}
    >
      Solid-white
    </Button>
    <Button
      variant="solid-blue"
      startIcon={<ViewIcon />}
      endIcon={<ArrowNextIcon />}
    >
      Solid-blue
    </Button>
    <Button
      variant="yellow"
      startIcon={<ViewIcon />}
      endIcon={<ArrowNextIcon />}
    >
      Yellow
    </Button>
    <Button
      variant="ghost"
      startIcon={<ViewIcon />}
      endIcon={<ArrowNextIcon />}
    >
      Ghost
    </Button>
    <Button
      variant="white-grey"
      startIcon={<ViewIcon />}
      endIcon={<ArrowNextIcon />}
    >
      White-grey
    </Button>
    <Button variant="link" startIcon={<ViewIcon />} endIcon={<ArrowNextIcon />}>
      Link
    </Button>
  </ButtonStack>
);

export const WithDifferentTag: Story<ButtonProps> = () => (
  <ButtonStack>
    <Button variant="solid-white" tag="a">
      Solid-white
    </Button>
    <Button variant="solid-blue" tag="a">
      Solid-blue
    </Button>
    <Button variant="yellow" tag="a">
      Yellow
    </Button>
    <Button variant="ghost" tag="a">
      Ghost
    </Button>
    <Button variant="white-grey" tag="a">
      White-grey
    </Button>
    <Button variant="link" tag="a">
      Link
    </Button>
  </ButtonStack>
);
