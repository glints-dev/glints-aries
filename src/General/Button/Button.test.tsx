import * as React from 'react';
import Button from '.';

import * as renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import { fireEvent, render } from '@testing-library/react';

import { ButtonVariantType, transformVariant } from './Button';
import { ButtonVariant, ButtonTheme } from '../../Utils/StyleConfig';

it('<Button> should render with text "click me" and an onClick handler', () => {
  const onClick = jest.fn();
  const ButtonSnapshot = renderer
    .create(<Button onClick={onClick}>click me</Button>)
    .toJSON();
  expect(ButtonSnapshot).toMatchSnapshot();
});

it('<Button> should render as button by default', () => {
  const { getByTestId } = render(
    <Button data-testid="test-button">test</Button>
  );
  expect(getByTestId('test-button').tagName).toBe('BUTTON');
});

describe('<Button> should render as the given tag', () => {
  const tags = ['a', 'h1', 'div'];
  tags.forEach((tag: string) => {
    it(`<Button> should render as ${tag}`, () => {
      const { container } = render(
        <Button tag={tag as React.ElementType}>button</Button>
      );
      const tagRegexp = new RegExp(tag, 'i');
      expect(container.querySelector(tag).tagName).toMatch(tagRegexp);
    });
  });
});

it('should call onClick once when it is clicked', () => {
  const onClick = jest.fn();
  const { getByText } = render(<Button onClick={onClick}>click me</Button>);
  const button = getByText('click me');
  fireEvent.click(button);
  expect(onClick).toHaveBeenCalledTimes(1);
});

it('should not call onClick when it is disabled', () => {
  const onClick = jest.fn();
  const { getByText } = render(
    <Button onClick={onClick} disabled>
      click me
    </Button>
  );
  const button = getByText('click me');
  fireEvent.click(button);
  expect(onClick).toHaveBeenCalledTimes(0);
});

describe('contains the correct className for each variant:', () => {
  const classNameMap = {
    [ButtonVariant.SOLID_BLUE]: 'aries-solid-btn',
    [ButtonVariant.SOLID_BLUE]: 'aries-solid-btn',
    [ButtonVariant.YELLOW]: 'aries-solid-shadow-btn',
    [ButtonVariant.GHOST]: 'aries-ghostbtn',
    [ButtonVariant.LINK]: 'aries-linkbtn',
  };

  Object.keys(classNameMap).forEach(variant => {
    it(`${variant}`, () => {
      const { getByText } = render(
        <Button variant={variant as ButtonVariantType}>click me</Button>
      );
      if (variant === ButtonVariant.LINK) {
        const linkButton = getByText('click me');
        expect(linkButton).toHaveClass(classNameMap[variant]);
      } else {
        const buttonContainer = getByText('click me').parentNode;
        expect(buttonContainer).toHaveClass(classNameMap[variant]);
      }
    });
  });
});

describe('transform the legacy variant and theme props to corresponding variant', () => {
  const deprecatedTheme = 'red';

  it('default', () => {
    expect(transformVariant('default')).toBe(ButtonVariant.SOLID_WHITE);
    expect(transformVariant('default', ButtonTheme.BLUE)).toBe(
      ButtonVariant.SOLID_BLUE
    );
    expect(transformVariant('default', deprecatedTheme)).toBe(
      ButtonVariant.SOLID_BLUE
    );
  });

  it('primary', () => {
    expect(transformVariant('primary')).toBe(ButtonVariant.YELLOW);
    expect(transformVariant('primary', ButtonTheme.BLUE)).toBe(
      ButtonVariant.YELLOW
    );
    expect(transformVariant('primary', deprecatedTheme)).toBe(
      ButtonVariant.YELLOW
    );
  });

  Object.values(ButtonVariant).forEach(variant => {
    it(`${variant}`, () => {
      expect(transformVariant(variant)).toBe(variant);

      if (variant === ButtonVariant.SOLID_WHITE) {
        // To check the legacy use of <Button theme="blue" />
        expect(transformVariant(variant, ButtonTheme.BLUE)).toBe(
          ButtonVariant.SOLID_BLUE
        );
        expect(transformVariant(variant, deprecatedTheme)).toBe(
          ButtonVariant.SOLID_BLUE
        );
      } else {
        expect(transformVariant(variant, ButtonTheme.BLUE)).toBe(variant);
        expect(transformVariant(variant, deprecatedTheme)).toBe(variant);
      }
    });
  });
});
