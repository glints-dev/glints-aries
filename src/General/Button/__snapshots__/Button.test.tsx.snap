// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button> should render with text "click me" and an onClick handler 1`] = `
<div
  className="ButtonStyle__SolidBtnContainer-jyb3o2-2 fqPreK aries-solid-btn"
  disabled={false}
>
  <button
    className="ButtonStyle__Button-jyb3o2-0 ButtonStyle__SolidBtn-jyb3o2-1 iLyOhD solid-btn-content"
    disabled={false}
    onClick={[MockFunction]}
  >
    click me
  </button>
</div>
`;
